### PR TITLE
upgrade mechanism

### PR DESCRIPTION
**Pull request by schubi2**

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -31,16 +31,96 @@ task :test, [:name] do |name, args|
...
     autoyast_file = File.join(base_dir, "kiwi/definitions/autoyast/autoinst.xml")
-    FileUtils.cp( File.join(base_dir, "spec", test_name + ".xml"), autoyast_file)
+    dest_definition = File.join(base_dir, "kiwi/definitions/autoyast/definition.rb")
+    default_iso = "http://dist.suse.de/install/SLE-12-Server-GM/SLE-12-Server-DVD-x86_64-GM-DVD1.iso"
+
+    FileUtils.rm(testing_iso, :force => true)
+
+    # Set used iso
+    unless test_name.start_with?("upgrade_")
+      # New installation workflow. So we are take the built OBS iso
```
Comment (imobach) [inline Rakefile]: `So we take the built OBS iso`

File: Rakefile [hunk 2]
```diff
@@ -31,16 +31,96 @@ task :test, [:name] do |name, args|
...
+      FileUtils.ln(obs_iso, testing_iso) if File.file?(obs_iso)
+    end
+    # Set download iso (if it is not already available)
```
Comment (imobach) [inline Rakefile]: This comment is confusing for me (maybe only for me :) )

File: Rakefile [hunk 3]
```diff
@@ -31,16 +31,96 @@ task :test, [:name] do |name, args|
...
+    end
+    system "sed -i.bak s,__iso_source_path__,#{data},g #{dest_definition}"
+
+    # Set autoinst.xml
+    autoinst = File.join(base_dir, "spec", test_name + ".xml")
+    autoinst = File.join(base_dir, "spec", test_name + ".install_xml") unless File.file?(autoinst)
+    FileUtils.cp( autoinst, autoyast_file)
+
+    puts "\n****** Creating KVM image ******\n"
     Dir.chdir(File.join( base_dir, "kiwi")) {
```
Comment (imobach) [inline Rakefile]: We must use `do...end` instead of curly braces.

File: Rakefile [hunk 4]
```diff
@@ -31,16 +31,96 @@ task :test, [:name] do |name, args|
...
+      # Take the generated image and run a autoyast upgrade workflow
+
+      # Set used iso
+      FileUtils.rm(testing_iso, :force => true)
+      FileUtils.ln(obs_iso, testing_iso) if File.file?(obs_iso) #Taking obs iso for upgrade
+
+      # Change boot order to CD on the top
+      system "sudo virsh destroy autoyast"
+      system "sudo virsh dumpxml autoyast >autoyast_description.xml"
+      system "sudo sed -i.bak s/dev=\\'cdrom\\'/dev=\\'cdrom_save\\'/g autoyast_description.xml"
```
Comment (imobach) [inline Rakefile]: Maybe I'm missing some detail, but I'd say that it should be enough changing `boot dev='hd'` for `boot dev='cdrom'`. Or not?

File: Rakefile [hunk 5]
```diff
@@ -31,16 +31,96 @@ task :test, [:name] do |name, args|
...
+
+      # Set autoinst.xml
+      autoinst = File.join(base_dir, "spec", test_name + ".upgrade_xml")
+      unless File.file?(autoinst)
+        puts "ERROR: #{autoinst} not found"
+        exit 1
+      end
+      FileUtils.cp( autoinst, autoyast_file)
+
+      Dir.chdir(File.join( base_dir, "kiwi")) {
```
Comment (imobach) [inline Rakefile]: `do...end`

File: Rakefile [hunk 6]
```diff
@@ -31,16 +31,96 @@ task :test, [:name] do |name, args|
...
+        puts "\n**** Updating KVM image ****\n"
+        system "veewee kvm build autoyast --force --auto"
+      }
+    end
+
+    Dir.chdir(File.join( base_dir, "kiwi")) {
```
Comment (imobach) [inline Rakefile]: `do...end`

**Comment by imobach** [comment]
In general, it looks good. However, the main rake task is getting bigger and bigger :) When everything is working, we should start thinking about some refactoring. But, first, let's make it work.

File: Rakefile (hunk 2 above)
Comment (schubi2) [inline Rakefile]: It is not you. It is me :-) I have changed.

File: Rakefile (hunk 4 above)
Comment (schubi2) [inline Rakefile]: not really. It is like exchanging values of two variables:
  <os>
    <type arch='x86_64' machine='pc-i440fx-2.1'>hvm</type>
    <boot dev='hd'/>
    <boot dev='cdrom'/>
  </os>

should become:
  <os>
    <type arch='x86_64' machine='pc-i440fx-2.1'>hvm</type>
    <boot dev='cdrom'/>
    <boot dev='hd'/>
    <boot dev='network'/>
  </os>

Comment (imobach) [inline Rakefile]: Ok, the file in the Pennyworth repository only contained one entry so I missed that point :+1:

**Comment by imobach** [comment]
LGTM

**Pull request merged**